test/pipelinerun_test.go

```diff
@@ -35,175 +36,288 @@ import (
+	for i, td := range tds {
+		t.Run(td.name, func(t *testing.T) {
+			logger.Infof("Setting up test resources for %q test in namespace %s", td.name, namespace)
+			td.testSetup(i)
 
-	if _, err := c.KubeClient.Kube.CoreV1().Secrets(namespace).Create(createPipelineRunSecret(namespace)); err != nil {
-		t.Fatalf("Failed to create secret `%s`: %s", hwSecret, err)
-	}
+			prName := fmt.Sprintf("%s%d", hwPipelineRunName, i)
+			if _, err := c.PipelineRunClient.Create(getHelloWorldPipelineRun(i, namespace)); err != nil {
+				t.Fatalf("Failed to create PipelineRun `%s`: %s", prName, err)
+			}
 
-	if _, err := c.KubeClient.Kube.CoreV1().ServiceAccounts(namespace).Create(createPipelineRunServiceAccount(namespace)); err != nil {
-		t.Fatalf("Failed to create SA `%s`: %s", hwSA, err)
-	}
+			logger.Infof("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+			if err := WaitForPipelineRunState(c, prName, func(tr *v1alpha1.PipelineRun) (bool, error) {
+				c := tr.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+				if c != nil {
+					if c.IsTrue() {
+						return true, nil
+					} else if c.IsFalse() {
+						return true, fmt.Errorf("Pipeline run has %s failed with status %v", prName, c.Status)
+					}
+				}
+				return false, nil
+			}, "PipelineRunSuccess"); err != nil {
+				t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+			}
+
+			logger.Infof("Making sure the expected TaskRuns %s were created", td.expectedTaskRuns)
+
+			expectedTaskRunNames := []string{}
+			for _, runName := range td.expectedTaskRuns {
+				taskRunName := strings.Join([]string{prName, runName}, "-")
+				expectedTaskRunNames = append(expectedTaskRunNames, taskRunName)
+				r, err := c.TaskRunClient.Get(taskRunName, metav1.GetOptions{})
+				if err != nil {
+					t.Fatalf("Couldn't get expected TaskRun %s: %s", taskRunName, err)
+				} else {
+					if !r.Status.GetCondition(duckv1alpha1.ConditionSucceeded).IsTrue() {
+						t.Fatalf("Expected TaskRun %s to have succeeded but Status is %v", taskRunName, r.Status)
+					}
+				}
+			}
+
+			matchKinds := map[string][]string{"PipelineRun": []string{prName}, "TaskRun": expectedTaskRunNames}
 
-	pp := getHelloWorldPipelineParams(namespace)
-	// Set SA pipeline params spec
-	pp.Spec.ServiceAccount = hwSA
-	if _, err := c.PipelineParamsClient.Create(pp); err != nil {
-		t.Fatalf("Failed to create PipelineParams `%s`: %s", hwPipelineParamsName, err)
+			logger.Infof("Making sure %d events were created from taskrun and pipelinerun with kinds %v", td.expectedNumberOfEvents, matchKinds)
+
+			events := collectMatchingEvents(t, c.KubeClient, namespace, matchKinds, "Succeeded")
+			if len(events) != td.expectedNumberOfEvents {
+				t.Fatalf("Expected %d number of successful events from pipelinerun and taskrun but got %d", td.expectedNumberOfEvents, len(events))
+			}
+			logger.Infof("Successfully finished test %q", td.name)
+		})
```

Comment:
this is excellent!! i really like it and I wouldn't have thought of it myself :D I love it! :heart: